### PR TITLE
Add LoggerBase and ITKLogger classes

### PR DESCRIPTION
**Pull request by blowekamp**
Add classes to provide the functionality of itk::OutputWindow to the
SimpleITK and deriving classes for custom behavior.

### CODE REVIEW

File: Code/Common/include/sitkLogger.h [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  DisplayDebugText(const char * t);
+
+  std::string
+  GetName() const override;
+  void
+  SetName(const std::string & n) override;
+
+  /** Set this logger instance to the ITK output window.
+   *
+   * After calling all ITK and SimpleITK display, warning and error texts will be handled by this object's methods.
```
Comment (dave3d) [inline Code/Common/include/sitkLogger.h:77]: 'After calling this method ...'

File: Code/Common/include/sitkLogger.h [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+class ITKLogger;
+
+/** \brief A base class to handle SimpleITK and ITK messages and logging.
+ *
+ * In ITK this function is implemented by the itk::OutputWindow, but in SimpleITK it is called a logger.
+ *
+ * Provides a base class for SimpleITK to provide the "DisplayText" methods that match the interface of
+ * itk::OutputWindow. Derived instances of LoggerBase are used by an internal adaptor derived from the itk::OutputWindow
+ * so that this object instances can be using in ITK.
```
Comment (zivy) [inline Code/Common/include/sitkLogger.h:44]: be using -> be used

File: Code/Common/include/sitkLogger.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+#  define SITK_FINAL
+#else
+#  define SITK_FINAL final
+#endif
+
+/** A class to hold itk::OutputWindows instances.
+ *
+ * All "DisplayText" methods are forwarded to the held OutputWindow instance.
+ *
+ * Neither the ownership nor callback functionality of the ObjectOwnedBase is not supported.
```
Comment (dave3d) [inline Code/Common/include/sitkLogger.h:103]: Neither is not supported?

File: Code/Common/include/sitkLogger.h [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+  void
+  DisplayGenericOutputText(const char * t) override;
+
+  void
+  DisplayDebugText(const char * t) override;
+
+  /** Set the ITK logger to the ITK OutputWindow instance held.
+   *
+   * Overrides the base functionality to elimicate the creation of the intermediary adaptor class.
```
Comment (dave3d) [inline Code/Common/include/sitkLogger.h:132]: eliminate

File: Code/Common/include/sitkLogger.h (hunk 3 above)
Comment (zivy) [inline Code/Common/include/sitkLogger.h:103]: Sentence starting with "Neither the ownership..." is unclear.

File: Testing/Unit/sitkCommonTests.cxx [hunk 5]
```diff
@@ -767,3 +768,102 @@ TEST( Kernel, Test1 )
...
+                                     "two\n"
+                                     "three\n"
+                                     "four\n"
+                                     "five\n";
+
+  EXPECT_EQ( testing::internal::GetCapturedStderr(), expectedOutput);
+}
+
+namespace {
+// A mockup of a logger which save messages to strings.
```
Comment (dave3d) [inline Testing/Unit/sitkCommonTests.cxx:801]: saves

File: Examples/Logging/Logging.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+import SimpleITK as sitk
+import logging
+
+
+class SimpleITKLogger(sitk.LoggerBase):
+    """
+     Adapts SimpleITK messages to be handled by a Python Logger object.
+
+     Allows using the logging module to control the handling of messages coming from ITK and SIMPLEITK. Messages
```
Comment (zivy) [inline Examples/Logging/Logging.py:27]: SIMPLEITK -> SimpleITK

File: Examples/Logging/Logging.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    def __enter__(self):
+        self._old_logger = self.SetAsGlobalITKLogger()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._old_logger.SetAsGlobalITKLogger()
+        del self._old_logger
+
+    def DisplayText(self, s):
+        # Remote newline endings from SimpleITK/ITK messages since the Python logger adds during output.
```
Comment (zivy) [inline Examples/Logging/Logging.py:65]: Remote -> Remove

**Review by zivy — APPROVED** [review]
Minor text comments. Very nice addition to SimpleITK. Logging is a feature that should be very useful for folks developing serious/commercial applications with SimpleITK.

**Review by zivy — APPROVED** [review]

**Review by dave3d — APPROVED** [review]
LGTM

**Pull request merged**